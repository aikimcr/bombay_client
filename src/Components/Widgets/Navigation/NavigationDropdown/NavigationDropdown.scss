.navigation-dropdown {
  background: transparent;

  &__button {
    font-size: var(--standard-button-font-size);
    background: transparent;
    border: none;
    cursor: pointer;
    anchor-name: --navigation-dropdown-button;
  }

  &__menu {
    background: var(--navigation-background-color);
    color: var(--navigation-color);
    margin: 0;
    position-anchor: --navigation-dropdown-button;
    position-area: y-end span-right;
    position-try: normal flip-start;
    border-radius: .5em;
  }
}