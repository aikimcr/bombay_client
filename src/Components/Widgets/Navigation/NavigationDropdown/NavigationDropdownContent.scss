.navigation-dropdown-content {
  display: grid;
  grid-auto-flow: row;
  gap: .5rem;
  background-color: transparent;
}