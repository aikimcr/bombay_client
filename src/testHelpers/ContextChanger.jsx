import React, { useState } from 'react';

import { BombayLoginContext } from '../Context/BombayLoginContext';
import { MemoryRouter } from 'react-router';

// Use this to turn the crank on context changes.
export const ContextChanger = ({ initialLoggedIn, children }) => {
  const [loggedIn, setLoggedIn] = useState(initialLoggedIn);

  const loginContext = {
    loggedIn,
    setLoggedIn,
  };

  function toggleLogin() {
    loginContext.setLoggedIn((oldLoggedIn) => !oldLoggedIn);
  }

  return (
    <BombayLoginContext.Provider value={loginContext}>
      <MemoryRouter>
        {children}
        <button
          className="change-test-login"
          data-testid="change-test-login"
          onClick={toggleLogin}
        >
          Change Login
        </button>
      </MemoryRouter>
    </BombayLoginContext.Provider>
  );
};
