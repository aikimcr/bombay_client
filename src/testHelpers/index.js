export * from './ContextChanger';
export * from './modelTools';
