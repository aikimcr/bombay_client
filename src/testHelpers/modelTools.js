// Import this after all mocks.  Especially Network and login mocks.

import casual from "casual";

import * as Network from "../Network/Network";

import ModelBase from "../Model/ModelBase";
import ArtistModel from "../Model/ArtistModel";
import SongModel from "../Model/SongModel";

const allNames = {};
casual.define("uniqueName", function (nameType, clear) {
  if (!allNames[nameType]) allNames[nameType] = new Set();

  const oldSize = allNames[nameType].size;
  let newName;

  while (allNames[nameType].size === oldSize) {
    newName = casual.full_name;
    allNames[nameType].add(newName);
  }

  return newName;
});

const idSequences = {};
casual.define("nextId", function (sequenceName) {
  const nextOne = (idSequences[sequenceName] ?? 0) + 1;
  idSequences[sequenceName] = nextOne;
  return nextOne;
});

export function makeAModel(tableName = "table1", fieldsCallback) {
  const def = {};
  def.id = casual.nextId(tableName);
  def.name = casual.uniqueName(tableName);

  if (fieldsCallback) {
    fieldsCallback(def);
  }

  let modelClass = ModelBase;

  switch (tableName) {
    case "artist":
      modelClass = ArtistModel;
      break;

    case "song":
      modelClass = SongModel;
      def.key_signature = "C";
      def.tempo = 120;
      def.lyrics = "O Solo Mio! The troubles I have seen";

      if (def.artist) {
        def.artist_id = def.artist.id;
      } else {
        const [artist] = makeAModel("artist");
        def.artist_id = artist.id;
        def.artist = artist;
      }
      break;

    default:
  }

  def.url = Network.buildURL({ path: `/${tableName}/${def.id}` });

  return [def, modelClass.from(def)];
}

export function makeModels(
  length = 10,
  query = {},
  tableName = "table1",
  fieldsCallback,
) {
  const result = {
    data: [],
  };

  const models = [];

  while (result.data.length < length) {
    const [def, model] = makeAModel(tableName, fieldsCallback);
    result.data.push(def);
    models.push(model);
  }

  let offset = (query.offset || 0) - length;
  let limit = query.limit || length;

  if (offset >= 0) {
    result.prevPage = Network.prepareURLFromArgs(tableName, {
      offset,
      limit,
    }).toString();
  }

  if (limit <= length) {
    offset = (query.offset || 0) + length;
    result.nextPage = Network.prepareURLFromArgs(tableName, {
      offset,
      limit,
    }).toString();
  }

  return [result, models];
}
