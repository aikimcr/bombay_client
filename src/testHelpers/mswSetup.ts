import { http, HttpResponse } from 'msw';
import {
  TestCollectionOneFetchBody,
  TestCollectionOneURL,
} from './TestCollectionTypes';

const TestData: TestCollectionOneFetchBody[] = [
  {
    data: [
      {
        id: 1,
        name: 'special voice melted',
        description: 'Ipsum amet consectetur commodo et ut et voluptate sint.',
      },
      {
        id: 2,
        name: 'picture than nails',
        description:
          'Reprehenderit aute ullamco quis qui duis do et quis minim labore esse.',
      },
      {
        id: 3,
        name: 'never ask dead',
        description:
          'Minim dolor eu fugiat tempor irure labore nostrud pariatur commodo aute.',
      },
      {
        id: 4,
        name: 'seems steady cry',
        description: 'Mollit voluptate nostrud nulla laborum nisi do occaecat.',
      },
      {
        id: 5,
        name: 'jack southern jet',
        description: 'Eiusmod qui culpa veniam eiusmod adipisicing aliqua.',
      },
      {
        id: 6,
        name: 'manner by trap',
        description:
          'Pariatur sit nulla laboris minim proident dolor aute amet id magna.',
      },
      {
        id: 7,
        name: 'perfect canal arrow',
        description:
          'Labore laboris magna reprehenderit aliquip ex Lorem sit non nisi exercitation ipsum voluptate.',
      },
      {
        id: 8,
        name: 'may river begun',
        description: 'Officia ipsum occaecat ad proident elit aliqua.',
      },
      {
        id: 9,
        name: 'donkey national writing',
        description:
          'Pariatur deserunt duis dolore velit cupidatat sunt ex duis duis veniam aliqua consectetur laborum.',
      },
      {
        id: 10,
        name: 'frog private would',
        description:
          'Commodo reprehenderit aute adipisicing esse nostrud pariatur incididunt eu.',
      },
    ],
    nextPage: `${TestCollectionOneURL}?offset=10&limit=10`,
  },
  {
    data: [
      {
        id: 11,
        name: 'notice visit happily',
        description: 'Ipsum amet consectetur commodo et ut et voluptate sint.',
      },
      {
        id: 12,
        name: 'condition aid hung',
        description:
          'Reprehenderit aute ullamco quis qui duis do et quis minim labore esse.',
      },
      {
        id: 13,
        name: 'variety class later',
        description:
          'Minim dolor eu fugiat tempor irure labore nostrud pariatur commodo aute.',
      },
      {
        id: 14,
        name: 'been edge student',
        description: 'Mollit voluptate nostrud nulla laborum nisi do occaecat.',
      },
      {
        id: 15,
        name: 'his brief length',
        description: 'Eiusmod qui culpa veniam eiusmod adipisicing aliqua.',
      },
      {
        id: 16,
        name: 'flame value fort',
        description:
          'Pariatur sit nulla laboris minim proident dolor aute amet id magna.',
      },
      {
        id: 17,
        name: 'police electric found',
        description:
          'Labore laboris magna reprehenderit aliquip ex Lorem sit non nisi exercitation ipsum voluptate.',
      },
      {
        id: 18,
        name: 'letter beneath day',
        description: 'Officia ipsum occaecat ad proident elit aliqua.',
      },
      {
        id: 19,
        name: 'mix leaving worse',
        description:
          'Pariatur deserunt duis dolore velit cupidatat sunt ex duis duis veniam aliqua consectetur laborum.',
      },
      {
        id: 20,
        name: 'younger simplest swept',
        description:
          'Commodo reprehenderit aute adipisicing esse nostrud pariatur incididunt eu.',
      },
    ],
    nextPage: `${TestCollectionOneURL}?offset=20&limit=10`,
    prevPage: `${TestCollectionOneURL}?offset=0&limit=10`,
  },
  {
    data: [
      {
        id: 21,
        name: 'barn mixture handle',
        description: 'Ipsum amet consectetur commodo et ut et voluptate sint.',
      },
      {
        id: 22,
        name: 'pair arm temperature',
        description:
          'Reprehenderit aute ullamco quis qui duis do et quis minim labore esse.',
      },
      {
        id: 23,
        name: 'whom answer try',
        description:
          'Minim dolor eu fugiat tempor irure labore nostrud pariatur commodo aute.',
      },
      {
        id: 24,
        name: 'there father bread',
        description: 'Mollit voluptate nostrud nulla laborum nisi do occaecat.',
      },
      {
        id: 25,
        name: 'laid correct lack',
        description: 'Eiusmod qui culpa veniam eiusmod adipisicing aliqua.',
      },
      {
        id: 26,
        name: 'factor willing dead',
        description:
          'Pariatur sit nulla laboris minim proident dolor aute amet id magna.',
      },
      {
        id: 27,
        name: 'date chair free',
        description:
          'Labore laboris magna reprehenderit aliquip ex Lorem sit non nisi exercitation ipsum voluptate.',
      },
      {
        id: 28,
        name: 'love mad nearest',
        description: 'Officia ipsum occaecat ad proident elit aliqua.',
      },
      {
        id: 29,
        name: 'truth eleven differ',
        description:
          'Pariatur deserunt duis dolore velit cupidatat sunt ex duis duis veniam aliqua consectetur laborum.',
      },
      {
        id: 30,
        name: 'younger describe eight',
        description:
          'Commodo reprehenderit aute adipisicing esse nostrud pariatur incididunt eu.',
      },
    ],
    prevPage: `${TestCollectionOneURL}?offset=10&limit=10`,
  },
];

export const TestCollectionHandlers = [
  http.get(TestCollectionOneURL, ({ request }) => {
    const requestUrl = new URL(request.url);

    if (requestUrl.searchParams.size > 0) {
      const offset = requestUrl.searchParams.get('offset');
      if (offset === '10') {
        return HttpResponse.json(TestData[1]);
      }

      if (offset === '20') {
        return HttpResponse.json(TestData[2]);
      }
    }

    return HttpResponse.json(TestData[0]);
  }),
];
