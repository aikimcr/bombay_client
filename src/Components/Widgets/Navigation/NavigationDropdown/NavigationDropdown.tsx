import React, {
  MouseEventHandler,
  ToggleEventHandler,
  useContext,
  useRef,
  useState,
} from 'react';
import BombayLoginContext from '../../../../Context/BombayLoginContext';
import { useRouteManager } from '../../../../Hooks/useRouteManager';
import { isHTMLElement } from '../../../../Utilities';

import './NavigationDropdown.scss';
import { NavigationDropdownContent } from './NavigationDropdownContent';

export const NavigationDropdown: React.FC = (): React.ReactElement => {
  const menuRef = useRef(null);

  const { navigateToRoute } = useRouteManager();
  const { loggedIn } = useContext(BombayLoginContext);
  const [isOpen, setIsOpen] = useState(false);

  const handleNavigationClick: MouseEventHandler = (evt) => {
    if (isHTMLElement(evt.target)) {
      navigateToRoute(evt.target.dataset.path);
    }

    menuRef.current?.hidePopover();
  };

  const handleToggle: ToggleEventHandler = (evt) => {
    setIsOpen(evt.newState === 'open');
  };

  if (!loggedIn) {
    return null;
  }

  return (
    <div className="navigation-dropdown">
      <button
        id="navigation-dropdown__button"
        className="navigation-dropdown__button"
        data-testid="navigation-dropdown-button"
        type="button"
        aria-haspopup="menu"
        aria-expanded={isOpen}
        aria-controls="navigation-dropdown__menu"
        popoverTarget="navigation-dropdown__menu"
        popoverTargetAction="toggle"
      >
        <i className="fa-solid fa-bars"></i>
      </button>
      <dialog
        ref={menuRef}
        id="navigation-dropdown__menu"
        className="navigation-dropdown__menu"
        data-testid="navigation-dropdown-menu"
        popover="auto"
        onToggle={handleToggle}
      >
        <NavigationDropdownContent
          handleNavigationClick={handleNavigationClick}
        />
      </dialog>
    </div>
  );
};
