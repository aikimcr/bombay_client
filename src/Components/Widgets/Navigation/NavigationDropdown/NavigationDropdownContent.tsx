import React, { MouseEventHandler } from 'react';
import { NavigationButton } from '..';

import './NavigationDropdownContent.scss';

interface NavigationDropdownContentProps {
  handleNavigationClick: MouseEventHandler;
}

export const NavigationDropdownContent: React.FC<
  NavigationDropdownContentProps
> = ({ handleNavigationClick }): React.ReactElement => {
  return (
    <div className="navigation-dropdown-content">
      <NavigationButton
        labelText="Artists"
        path="/artistList"
        onClick={handleNavigationClick}
      />
      <NavigationButton
        labelText="Songs"
        path="/songList"
        onClick={handleNavigationClick}
      />
    </div>
  );
};
