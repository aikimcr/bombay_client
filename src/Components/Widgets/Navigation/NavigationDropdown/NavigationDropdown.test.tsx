import React from 'react';
import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import {
  mockLoginStatus,
  mockRefreshToken,
  mockLogin,
  mockLogout,
} from '../../../../Network/testing';

jest.mock('../../../../Network/Login', () => {
  const originalModule = jest.requireActual('../../../../Network/Login');

  return {
    __esModule: true,
    ...originalModule,
    loginStatus: mockLoginStatus,
    refreshToken: mockRefreshToken,
    login: mockLogin,
    logout: mockLogout,
  };
});

import {
  mockUseRouteManager,
  mockNavigateToRoute,
} from '../../../../Hooks/testing/mocks/mockUseRouteManager';

jest.mock('../../../../Hooks/useRouteManager', () => {
  const originalModule = jest.requireActual(
    '../../../../Hooks/useRouteManager',
  );

  return {
    __esModules: true,
    ...originalModule,
    useRouteManager: mockUseRouteManager,
  };
});

import { ContextChanger } from '../../../../testHelpers/ContextChanger';

import { NavigationDropdown } from './NavigationDropdown';

describe('Test NavigationDropdown', () => {
  describe('Logged In', () => {
    it('should match snapshot', () => {
      const { asFragment } = render(
        <ContextChanger initialLoggedIn={true}>
          <NavigationDropdown />
        </ContextChanger>,
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('should show the dropdown button', () => {
      render(
        <ContextChanger initialLoggedIn={true}>
          <NavigationDropdown />
        </ContextChanger>,
      );

      const button = screen.getByTestId('navigation-dropdown-button');
      expect(button).toBeInTheDocument();
      expect(button).toHaveAttribute('id', 'navigation-dropdown__button');
      expect(button).toHaveAttribute('aria-haspopup', 'menu');
      expect(button).toHaveAttribute('aria-expanded', 'false');
      expect(button).toHaveAttribute(
        'aria-controls',
        'navigation-dropdown__menu',
      );

      expect(button).toHaveClass('navigation-dropdown__button');
      expect(button).toHaveTextContent('');

      const fontAwesome = button.getElementsByTagName('i');
      expect(fontAwesome).toHaveLength(1);
      expect(fontAwesome[0]).toHaveClass('fa-solid');
      expect(fontAwesome[0]).toHaveClass('fa-bars');
    });

    it('should open the popover when the button is clicked', () => {
      render(
        <ContextChanger initialLoggedIn={true}>
          <NavigationDropdown />
        </ContextChanger>,
      );

      const button = screen.getByTestId('navigation-dropdown-button');
      expect(button).toHaveAttribute('aria-expanded', 'false');

      const dialog: HTMLDialogElement = screen.getByTestId(
        'navigation-dropdown-menu',
      );
      expect(dialog).not.toBeVisible();

      act(() => {
        fireEvent.click(button);
      });

      waitFor(() => {
        expect(dialog).toBeVisible();
      });

      waitFor(() => {
        expect(button).toHaveAttribute('aria-expanded', 'true');
      });
    });

    it('should contain the anchor list', () => {
      render(
        <ContextChanger initialLoggedIn={true}>
          <NavigationDropdown />
        </ContextChanger>,
      );

      const dialog: HTMLDialogElement = screen.getByTestId(
        'navigation-dropdown-menu',
      );
      const navItems = dialog.querySelectorAll('.navigation-button');
      expect(navItems).toHaveLength(2);
      expect(navItems[0]).toHaveAttribute('data-path', '/artistList');
      expect(navItems[1]).toHaveAttribute('data-path', '/songList');
    });

    it('should close the popover when an item is clicked', () => {
      const mockHidePopover = jest.fn();

      if (!HTMLElement.prototype.hidePopover) {
        HTMLElement.prototype.hidePopover = mockHidePopover;
      }

      render(
        <ContextChanger initialLoggedIn={true}>
          <NavigationDropdown />
        </ContextChanger>,
      );

      const dialog: HTMLDialogElement = screen.getByTestId(
        'navigation-dropdown-menu',
      );
      const navItem = dialog.querySelector(
        '.navigation-button[data-path="/artistList"]',
      );

      act(() => {
        fireEvent.click(navItem);
      });

      expect(mockNavigateToRoute).toHaveBeenCalledWith('/artistList');
    });
  });

  describe('logged out', () => {
    it('should be empty', () => {
      render(
        <ContextChanger initialLoggedIn={false}>
          <NavigationDropdown />
        </ContextChanger>,
      );

      expect(screen.queryByTestId('navigation-dropdown-button')).toBeNull();
      expect(screen.queryByTestId('navigation-dropdown-menu')).toBeNull();
    });
  });
});
