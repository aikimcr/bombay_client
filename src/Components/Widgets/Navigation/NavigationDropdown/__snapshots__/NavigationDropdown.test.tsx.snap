// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Test NavigationDropdown Logged In should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="navigation-dropdown"
  >
    <button
      aria-controls="navigation-dropdown__menu"
      aria-expanded="false"
      aria-haspopup="menu"
      class="navigation-dropdown__button"
      data-testid="navigation-dropdown-button"
      id="navigation-dropdown__button"
      popovertarget="navigation-dropdown__menu"
      popovertargetaction="toggle"
      type="button"
    >
      <i
        class="fa-solid fa-bars"
      />
    </button>
    <dialog
      class="navigation-dropdown__menu"
      data-testid="navigation-dropdown-menu"
      id="navigation-dropdown__menu"
      popover="auto"
    >
      <div
        class="navigation-dropdown-content"
      >
        <button
          class="navigation-button"
          data-path="/artistList"
          data-testid="navigation-button"
          type="button"
        >
          Artists
        </button>
        <button
          class="navigation-button"
          data-path="/songList"
          data-testid="navigation-button"
          type="button"
        >
          Songs
        </button>
      </div>
    </dialog>
  </div>
  <button
    class="change-test-login"
    data-testid="change-test-login"
  >
    Change Login
  </button>
</DocumentFragment>
`;
